feat(curriculum): add express middleware lessons (#67194)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/blocks/lecture-express-middleware/69f42cd936da600a74a84926.md

---
id: 69f42cd936da600a74a84926
title: What is Middleware and How Does it Work?
challengeType: 19
dashedName: what-is-middleware-and-how-does-it-work
---

# --description--

> Middleware is an essential concept in Express, and it allows us to handle requests and responses between the client and server. Middleware functions have access to the request object `(req)`, the response object `(res)`, and the next function in the application's request-response cycle.

Reviewer: ```suggestion
Middleware is an essential concept in Express, and it allows us to handle requests and responses between the client and server. Middleware functions have access to the request object, the response object, and the next function in the application's request-response cycle.
```

Those are just arbitrary parameter names.

Let's break it down:

## What is Middleware?

Middleware is essentially a function that executes during the lifecycle of a request to the server. It performs various tasks such as logging requests, handling errors, parsing incoming requests, or serving static files.

Express provides a convenient way to define middleware that processes incoming requests before they reach route handlers.

How Middleware Works:

When a request is made to an Express application, middleware is executed in the order it is added. The flow looks like this:

* Request is sent to the server.
* Middleware functions process the request.

> * If the middleware calls next(), it passes control to the next middleware or route handler.

Reviewer: ```suggestion
* If the middleware calls `next()`, it passes control to the next middleware or route handler.
```

> * If the middleware doesn't call next(), the request-response cycle stops.

Reviewer: ```suggestion
* If the middleware doesn't call `next()`, the request-response cycle stops.
```

```js
const express = require('express')
const app = express()

// Simple middleware that logs request info
app.use((req, res, next) => {
  console.log('Request URL:', req.url)
  console.log('Request Method:', req.method)
  next() // Passes control to the next middleware
})

app.get('/', (req, res) => {
  res.send('Hello World!')
})

app.listen(3000, () => {
  console.log('Server running on http://localhost:3000')
})
```

`app.use()`: Adds middleware globally to your application.

> The middleware logs request details and calls `next()` to proceed to the next middleware or route handler.

Reviewer: ```suggestion
In this example, the middleware logs request details and calls `next()` to proceed to the next middleware or route handler.
```

> Middleware is a powerful feature in Express that allows you to intercept and modify requests at various points. You can use middleware to handle things like logging, authentication, or even modify request and response objects before reaching route handlers.

Reviewer: ```suggestion
Middleware is a powerful feature in Express that allows you to intercept and modify requests at various points. You can use middleware to handle things like logging, authentication, or even modify request and response objects before they reach your route handlers.
```

# --questions--

## --text--

What does middleware in Express allow you to do?

## --answers--

> Serve static files and other assets.

Reviewer 2: This answer is marked wrong, but here the lesson says that middleware is used to serve static files:

https://github.com/Ksound22/freeCodeCamp/blob/b4fb3a15ca64db576e11b323eb6f5bf7fe2a3a60/curriculum/challenges/english/blocks/lecture-express-middleware/69f42cd936da600a74a84926.md?plain=1#L16

### --feedback--

> Think about which function executes during the lifecycle of a request to the server.

Reviewer: ```suggestion
Think about which functions execute during the lifecycle of a request to the server.
```

Multiple functions execute (the route handler is a function too~)

---

Modify the request and response objects.

---

Handle routes.

### --feedback--

> Think about which function executes during the lifecycle of a request to the server.

Reviewer: ```suggestion
Think about which functions execute during the lifecycle of a request to the server.
```

---

Start the server.

### --feedback--

> Think about which function executes during the lifecycle of a request to the server.

Reviewer: ```suggestion
Think about which functions execute during the lifecycle of a request to the server.
```

## --video-solution--

2

## --text--

What happens if middleware does not call `next()`?

## --answers--

> The request will continue through the next middleware.

Reviewer: ```suggestion
The request will continue through to the next middleware.
```

### --feedback--

> Think about the four steps involved in the activity of a middleware.

Reviewer: ```suggestion
Think about the four steps involved in the lifecycle of a request.
```

---

The request-response cycle will stop.

---

An error will be thrown.

### --feedback--

> Think about the four steps involved in the activity of a middleware.

Reviewer: ```suggestion
Think about the four steps involved in the lifecycle of a request.
```

---

The request will be handled by the last route handler.

### --feedback--

> Think about the four steps involved in the activity of a middleware.

Reviewer: ```suggestion
Think about the four steps involved in the lifecycle of a request.
```

## --video-solution--

2

## --text--

Which of the following is true about middleware?

## --answers--

Middleware runs only after a route handler is executed.

### --feedback--

> Think about how Middleware can inspect, change, or pass along requests and responses.

Reviewer: ```suggestion
Think about how middleware can inspect, change, or pass along requests and responses.
```

---

Middleware can modify the request and response objects.

---

Middleware can only log requests.

### --feedback--

> Think about how Middleware can inspect, change, or pass along requests and responses.

Reviewer: ```suggestion
Think about how middleware can inspect, change, or pass along requests and responses.
```

---

> Middleware only works with POST requests.

Reviewer: ```suggestion
Middleware only works with `POST` requests.
```

### --feedback--

> Think about how Middleware can inspect, change, or pass along requests and responses.

Reviewer: ```suggestion
Think about how middleware can inspect, change, or pass along requests and responses.
```
